# SpendDNA — UPI Transaction Analyzer
**Name:** Samrudhi Shetty | **Batch:** Data Science August 2026 | **Date:** September 1, 2026

Analyzing `rahul_transactions.csv` — synthetic UPI/bank transactions, 6 months, 1 user (Rahul Sharma).

Built using only Python fundamentals + NumPy + Pandas. No matplotlib, no regex, no Counter, no ML/AI libraries.

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("rahul_transactions.csv")

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1328 entries, 0 to 1327
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         1328 non-null   object 
 1   Time         1328 non-null   object 
 2   Description  1328 non-null   object 
 3   Type         1328 non-null   object 
 4   Amount       1328 non-null   object 
 5   Balance      1328 non-null   float64
 6   Mode         1328 non-null   object 
 7   Ref          1328 non-null   object 
dtypes: float64(1), object(7)
memory usage: 83.1+ KB


In [ ]:
df.shape

(1328, 8)

In [ ]:
df.describe(include="all")

,Date,Time,Description,Type,Amount,Balance,Mode,Ref
count,1328,1328,1328,1328,1328,1328.000000,1328,1328
unique,617,832,283,3,1141,NaN,4,1307
top,2024-02-07,16:19,POS SWIGGY BANGALORE,DR,"Rs. 15,000",NaN,UPI,TXN427270
freq,6,6,30,670,5,NaN,1299,2
mean,NaN,NaN,NaN,NaN,NaN,147805.330572,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,341059.809969,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,-769127.000000,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,-104585.750000,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,151650.000000,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,448324.250000,NaN,NaN


In [ ]:
df.columns

Index(['Date', 'Time', 'Description', 'Type', 'Amount', 'Balance', 'Mode',
       'Ref'],
      dtype='object')

In [ ]:
df.tail(15)

,Date,Time,Description,Type,Amount,Balance,Mode,Ref
1313,28 Jun 2024,21:33,BUNDL Tech P L,DR,₹589,-307229.0,UPI,TXN356161
1314,2024-06-28,21:54,BLINKIT BANGALORE,Debit,₹878,-305746.0,UPI,TXN974187
1315,29-Jun-24,00:12,UPI-OLACABS@HDFCBANK,Debit,215.00,-307614.0,UPI,TXN802374
1316,2024-06-29,10:35,UPI-BLINKIT-1828@PAYTM,DR,Rs. 216,-308860.0,UPI,TXN874494
1317,29-Jun-24,14:22,KIRANAKART TECH,DR,₹515,-310230.0,UPI,TXN962258
1318,29-Jun-24,16:39,UPI-RAPIDO@OKAXIS,Debit,170.00,-307399.0,UPI,TXN540114
1319,2024-06-29,18:36,POS BANGALORE RESTAURANT,Debit,"1,030.00",-308644.0,UPI,TXN658352
1320,29 Jun 2024,19:35,POS SWIGGY BANGALORE,DR,Rs. 405,-309265.0,UPI,TXN394032
1321,2024-06-29,20:46,ZOMATO-DINING,DR,₹203,-309715.0,UPI,TXN703638
1322,29/06/24,21:57,UPI-UBER-3218@HDFCBANK,DR,247.00,-309512.0,UPI,TXN667358


In [ ]:
df["Description"].unique()

array(['AMAZON SELLER SVCS', 'BHIM-BMTC',
       'NEFT-TECHCRUSH LABS-SALARY MAY24', 'UPI-AMAN-8934@OKAXIS',
       'BHIM-BLINKIT', 'BHIM ZEPTO', 'UPI-UBER-2426@HDFCBANK',
       'POS SWIGGY BANGALORE', 'UPI-GROWWPAY@HDFCBANK', 'OLA ELECTRIC',
       'BMS MOVIE TICKETS', 'POS OLA-PRIME', 'SWIGGY-INSTAMART',
       'UPI-STARBUCKS@AXIS', 'UPI-THIRDWAVE@OKAXIS', 'ANI Technologies',
       'BMTC BUS PASS', 'POS TRUFFLES', 'FLIPKART INDIA',
       'POS SWIGGY-RESTAURANT', 'GROFERS INDIA P L', 'POS UBER BANGALORE',
       'BANGALORE ELEC SUPPLY', 'TWC INDIA', 'UPI-BESCOM-BILL@HDFCBANK',
       'UPI-AMAN-0816@OKAXIS', 'ROPPEN TRANSPORTATION', 'OLA CABS',
       'POS ZOMATO', 'UPI-AMAZONPAY@HDFCBANK', 'POS BLINKIT',
       'IMPS-RENT-LANDLORD-75500265', 'ZOMATO MEDIA P L',
       'UPI-ANKIT-6430@OKAXIS', 'UPI-OLACABS@HDFCBANK',
       'UPI-JIORECHARGE@PAYTM', 'UPI-CCD@HDFCBANK', 'Swiggy*Order',
       'INSTAMART BANGALORE', 'UPI-ZOMATO-LIMITED@PAYTM',
       'AVENUE SUPERMARTS', 'POS HP PETROL

## Feature 1 - The Transaction Parser

In [ ]:
df_clean = df.copy()

In [ ]:
df_clean.isnull().sum()

,0
Date,0
Time,0
Description,0
Type,0
Amount,0
Balance,0
Mode,0
Ref,0


In [ ]:
df_clean.duplicated().sum()

np.int64(18)

In [ ]:
df_clean.drop_duplicates(inplace=True)

In [ ]:
df_clean.shape

(1310, 8)

In [ ]:
df_clean["Date"].head(20)

,Date
0,2024-01-01
1,01-Jan-24
2,01-Jan-24
3,2024-01-01
4,01 Jan 2024
5,2024-01-01
6,2024-01-01
7,01-Jan-24
8,2024-01-02
9,02/01/24


In [ ]:
df_clean["Date"] = pd.to_datetime(
    df_clean["Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

In [ ]:
df_clean["Date"].isnull().sum()

np.int64(0)

In [ ]:
df_clean["Date"].head(20)

,Date
0,2024-01-01
1,2024-01-01
2,2024-01-01
3,2024-01-01
4,2024-01-01
5,2024-01-01
6,2024-01-01
7,2024-01-01
8,2024-01-02
9,2024-01-02


In [ ]:
df_clean["Amount"].head(20)

,Amount
0,₹2462
1,50.00
2,₹84728
3,₹1828
4,270.00
5,Rs. 625
6,Rs. 148
7,₹482
8,Rs. 537
9,3956


In [ ]:
df_clean["Amount"] = (
    df_clean["Amount"]
    .astype(str)
    .str.replace("₹", "", regex=False)
    .str.replace("Rs.", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.strip()
)

In [ ]:
df_clean["Amount"] = pd.to_numeric(df_clean["Amount"], errors="coerce")

In [ ]:
print(df_clean["Amount"].dtype)
print(df_clean["Amount"].isnull().sum())
print(df_clean["Amount"].head())

float64
0
0     2462.0
1       50.0
2    84728.0
3     1828.0
4      270.0
Name: Amount, dtype: float64


In [ ]:
df_clean["Balance"].head(20)

,Balance
0,678275.0
1,681007.0
2,484728.0
3,-748745.0
4,680737.0
5,677650.0
6,677020.0
7,677168.0
8,676177.0
9,672221.0


In [ ]:
df_clean["Balance"] = (
    df_clean["Balance"]
    .astype(str)
    .str.replace("₹", "", regex=False)
    .str.replace("Rs.", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.strip()
)

df_clean["Balance"] = pd.to_numeric(df_clean["Balance"], errors="coerce")

In [ ]:
print(df_clean["Balance"].dtype)
print(df_clean["Balance"].isnull().sum())
print(df_clean["Balance"].head())

float64
0
0    678275.0
1    681007.0
2    484728.0
3   -748745.0
4    680737.0
Name: Balance, dtype: float64


In [ ]:
df_clean["Type"].unique()

array(['Debit', 'DR', 'CR'], dtype=object)

In [ ]:
df_clean["Type"] = df_clean["Type"].str.lower()

df_clean["Type"] = df_clean["Type"].replace({
    "dr": "debit",
    "cr": "credit"
})

In [ ]:
df_clean["Type"].unique()

array(['debit', 'credit'], dtype=object)

In [ ]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1310 entries, 0 to 1327
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         1310 non-null   datetime64[ns]
 1   Time         1310 non-null   object        
 2   Description  1310 non-null   object        
 3   Type         1310 non-null   object        
 4   Amount       1310 non-null   float64       
 5   Balance      1310 non-null   float64       
 6   Mode         1310 non-null   object        
 7   Ref          1310 non-null   object        
dtypes: datetime64[ns](1), float64(2), object(5)
memory usage: 92.1+ KB


In [ ]:
df_clean["Type"].unique()

array(['debit', 'credit'], dtype=object)

In [ ]:
df_clean["Mode"] = df_clean["Mode"].replace("", np.nan)



---



## Feature 2 - Vendor Extractor

In [ ]:
for desc in df_clean["Description"].unique()[:100]:
    print(desc)

AMAZON SELLER SVCS
BHIM-BMTC
NEFT-TECHCRUSH LABS-SALARY MAY24
UPI-AMAN-8934@OKAXIS
BHIM-BLINKIT
BHIM ZEPTO
UPI-UBER-2426@HDFCBANK
POS SWIGGY BANGALORE
UPI-GROWWPAY@HDFCBANK
OLA ELECTRIC
BMS MOVIE TICKETS
POS OLA-PRIME
SWIGGY-INSTAMART
UPI-STARBUCKS@AXIS
UPI-THIRDWAVE@OKAXIS
ANI Technologies
BMTC BUS PASS
POS TRUFFLES
FLIPKART INDIA
POS SWIGGY-RESTAURANT
GROFERS INDIA P L
POS UBER BANGALORE
BANGALORE ELEC SUPPLY
TWC INDIA
UPI-BESCOM-BILL@HDFCBANK
UPI-AMAN-0816@OKAXIS
ROPPEN TRANSPORTATION
OLA CABS
POS ZOMATO
UPI-AMAZONPAY@HDFCBANK
POS BLINKIT
IMPS-RENT-LANDLORD-75500265
ZOMATO MEDIA P L
UPI-ANKIT-6430@OKAXIS
UPI-OLACABS@HDFCBANK
UPI-JIORECHARGE@PAYTM
UPI-CCD@HDFCBANK
Swiggy*Order
INSTAMART BANGALORE
UPI-ZOMATO-LIMITED@PAYTM
AVENUE SUPERMARTS
POS HP PETROL STATION
UPI-VIKAS-6060@OKAXIS
POS BANGALORE RESTAURANT
UPI-ZERODHA-COIN@AXIS
BHIM SWIGGY
UPI-BOOKMYSHOW@HDFCBANK
BLINKIT BANGALORE
POS ZEPTO BANGALORE
UPI-DMART@OKAXIS
UPI-RAPIDO@OKAXIS
BUNDL Tech P L
TATA STARBUCKS
POS CAFE COFFEE DAY

In [ ]:
vendor_keywords = {
    "Swiggy": ["SWIGGY", "BUNDL", "INSTAMART"],
    "Zomato": ["ZOMATO"],
    "Blinkit": ["BLINKIT"],
    "Zepto": ["ZEPTO"],
    "Amazon": ["AMAZON", "AMZN"],
    "Flipkart": ["FLIPKART", "FKART"],
    "Myntra": ["MYNTRA"],
    "Nykaa": ["NYKAA", "FSN", "E-COMMERCE"],
    "Uber": ["UBER"],
    "Ola": ["OLA", "ANI TECHNOLOGIES"],
    "Rapido": ["RAPIDO", "ROPPEN TRANSPORTATION"],
    "BMTC": ["BMTC"],
    "Starbucks": ["STARBUCKS"],
    "Third Wave Coffee": ["THIRDWAVE", "TWC", "THIRD WAVE COFFEE"],
    "Cafe Coffee Day": ["CCD", "COFFEE DAY"],
    "BookMyShow": ["BOOKMYSHOW", "BMS", "BIGTREE ENTERTAINMENT"],
    "BigBasket": ["BIGBASKET", "KIRANAKART TECH"],
    "DMart": ["DMART", "AVENUE SUPERMARKETS", "AVENUE SUPERMARTS"],
    "Innovative Retail": ["INNOVATIVE RETAIL"],
    "Groww": ["GROWW"],
    "Zerodha": ["ZERODHA"],
    "Netflix": ["NETFLIX"],
    "Spotify": ["SPOTIFY"],
    "Disney+ Hotstar": ["HOTSTAR", "DISNEY"],
    "Airtel": ["AIRTEL"],
    "Jio": ["JIO"],
    "Vi": ["VI POSTPAID"],
    "BESCOM": ["BESCOM", "BANGALORE ELEC SUPPLY"],
    "BWSSB": ["BWSSB"],
    "Indian Oil": ["IOC", "INDIAN OIL"],
    "HP Petrol": ["HP PETROL"],
    "BPCL": ["BPCL"]
}

In [ ]:
def extract_vendor(description):
    description = str(description).upper()

    # P2P transfers
    if ("UPI-AMAN" in description or
        "UPI-ANKIT" in description or
        "UPI-VIKAS" in description or
        "UPI-PRIYA" in description):
        return "P2P Transfer"

    # ATM withdrawals
    if "ATM" in description and "WDL" in description:
        return "Cash Withdrawal"

    # Vendor matching
    for vendor, keywords in vendor_keywords.items():
        for keyword in keywords:
            if keyword in description:
                return vendor

    return "Uncategorised"

In [ ]:
df_clean["Vendor"] = df_clean["Description"].apply(extract_vendor)

In [ ]:
df_clean[["Description", "Vendor"]].head(20)

,Description,Vendor
0,AMAZON SELLER SVCS,Amazon
1,BHIM-BMTC,BMTC
2,NEFT-TECHCRUSH LABS-SALARY MAY24,Uncategorised
3,UPI-AMAN-8934@OKAXIS,P2P Transfer
4,BHIM-BLINKIT,Blinkit
5,BHIM ZEPTO,Zepto
6,UPI-UBER-2426@HDFCBANK,Uber
7,BHIM-BLINKIT,Blinkit
8,POS SWIGGY BANGALORE,Swiggy
9,UPI-GROWWPAY@HDFCBANK,Groww


In [ ]:
df_clean["Vendor"].value_counts()

,count
Vendor,
Swiggy,243
Zomato,121
Uncategorised,112
Ola,87
Amazon,86
Uber,71
Zepto,58
Rapido,55
Flipkart,47


In [ ]:
df_clean[df_clean["Vendor"] == "Others"][["Description"]].drop_duplicates()

,Description


In [ ]:
vendor_keywords.update({
    "Salary": ["SALARY"],
    "Person Transfer": ["AMAN", "ANKIT", "PRIYA", "VIKAS"],
    "Restaurant": ["RESTAURANT", "TRUFFLES", "MEGHANA", "EMPIRE"],
    "ATM Withdrawal": ["ATM-WDL"],
    "Grofers": ["GROFERS"]
})

In [ ]:
df_clean["Vendor"] = df_clean["Description"].apply(extract_vendor)

In [ ]:
df_clean[df_clean["Vendor"] == "Others"][["Description"]].drop_duplicates()

,Description


In [ ]:
df_clean["Vendor"].value_counts()

,count
Vendor,
Swiggy,243
Zomato,121
Ola,87
Amazon,86
Uber,71
Restaurant,63
Zepto,58
Rapido,55
Flipkart,47


In [ ]:
df_clean[df_clean["Vendor"] == "Others"]["Description"].drop_duplicates().sort_values().tolist()

[]

In [ ]:
vendor_keywords.update({
    "BookMyShow": ["BIGTREE ENTERTAINMENT", "BOOKMYSHOW", "BMS"],
    "Nykaa": ["FSN E-COMMERCE", "NYKAA", "FSN"],
    "Rent": ["LANDLORD", "RENT"],
    "BigBasket": ["KIRANAKART TECH", "BIGBASKET"],
    "Dineout": ["DINEOUT"],
    "Third Wave Coffee": ["THIRD WAVE COFFEE", "THIRDWAVE", "TWC"],
    "Rapido": ["ROPPEN TRANSPORTATION", "RAPIDO"],
    "Disney+ Hotstar": ["STAR INDIA PVT LTD", "HOTSTAR", "DISNEY"],
    "Person Transfer": ["KARAN", "NEHA", "SNEHA"],
    "Vi": ["VODAFONE IDEA", "VI-RECHARGE", "VI POSTPAID"]
})

In [ ]:
df_clean["Vendor"] = df_clean["Description"].apply(extract_vendor)

In [ ]:
df_clean["vendor_clean"] = df_clean["Description"].apply(extract_vendor)

print("Unique vendors:", df_clean["vendor_clean"].nunique())

print("\nTop 10 vendors:")
print(df_clean["vendor_clean"].value_counts().head(10))

print("\nUncategorised descriptions:")
print(df_clean[df_clean["vendor_clean"] == "Uncategorised"]["Description"].unique())

print("\nNumber of uncategorised:",
      (df_clean["vendor_clean"] == "Uncategorised").sum())

Unique vendors: 40

Top 10 vendors:
vendor_clean
Swiggy        243
Zomato        121
Ola            87
Amazon         86
Uber           71
Restaurant     63
Zepto          58
Rapido         55
Flipkart       47
Starbucks      42
Name: count, dtype: int64

Uncategorised descriptions:
[]

Number of uncategorised: 0




---



# Feature 3 - Category Tagger

In [ ]:
category_map = {
    "Swiggy": "Food Delivery",
    "Zomato": "Food Delivery",

    "Blinkit": "Quick Commerce",
    "Zepto": "Quick Commerce",
    "Grofers": "Quick Commerce",

    "Amazon": "E-commerce",
    "Flipkart": "E-commerce",
    "Myntra": "E-commerce",
    "Nykaa": "E-commerce",

    "Uber": "Transport",
    "Ola": "Transport",
    "Rapido": "Transport",
    "BMTC": "Transport",

    "Starbucks": "Cafe",
    "Third Wave Coffee": "Cafe",
    "Cafe Coffee Day": "Cafe",

    "Restaurant": "Restaurants",
    "Dineout": "Restaurants",

    "Netflix": "Subscriptions",
    "Spotify": "Subscriptions",
    "Amazon Prime": "Subscriptions",
    "Disney+ Hotstar": "Subscriptions",

    "Airtel": "Utilities",
    "Jio": "Utilities",
    "Vi": "Utilities",
    "BESCOM": "Utilities",
    "BWSSB": "Utilities",

    "BigBasket": "Groceries",
    "DMart": "Groceries",
    "Innovative Retail": "Groceries",

    "Groww": "Investments",
    "Zerodha": "Investments",

    "Indian Oil": "Fuel",
    "HP Petrol": "Fuel",
    "BPCL": "Fuel",

    "BookMyShow": "Entertainment",

    "Salary": "Salary",
    "Rent": "Rent",

    "P2P Transfer": "Personal Transfer",
    "Person Transfer": "Personal Transfer",
    "Cash Withdrawal": "Cash Withdrawal"
}

In [ ]:
df_clean["category"] = df_clean["vendor_clean"].map(category_map)

In [ ]:
df_clean["category"] = df_clean["vendor_clean"].map(category_map)

In [ ]:
df_clean[["vendor_clean", "category"]].head(20)

,vendor_clean,category
0,Amazon,E-commerce
1,BMTC,Transport
2,Salary,Salary
3,P2P Transfer,Personal Transfer
4,Blinkit,Quick Commerce
5,Zepto,Quick Commerce
6,Uber,Transport
7,Blinkit,Quick Commerce
8,Swiggy,Food Delivery
9,Groww,Investments


In [ ]:
df_clean["category"].value_counts()

,count
category,
Food Delivery,364
Transport,250
E-commerce,172
Quick Commerce,113
Cafe,99
Restaurants,73
Groceries,54
Utilities,43
Subscriptions,31


In [ ]:
print("Uncategorised transactions:", df_clean["category"].isna().sum())

print("\nCategory counts:")
print(df_clean["category"].value_counts())

Uncategorised transactions: 0

Category counts:
category
Food Delivery        364
Transport            250
E-commerce           172
Quick Commerce       113
Cafe                  99
Restaurants           73
Groceries             54
Utilities             43
Subscriptions         31
Fuel                  28
Investments           23
Personal Transfer     18
Cash Withdrawal       17
Entertainment         13
Salary                 6
Rent                   6
Name: count, dtype: int64




---



# Feature 4 - Spending Overview


In [ ]:
# Total Income
total_income = df_clean[df_clean["Type"] == "credit"]["Amount"].sum()
print("Total Income: ₹", total_income)

Total Income: ₹ 509774.0


In [ ]:
#Total Expense
total_expense = df_clean[df_clean["Type"] == "debit"]["Amount"].sum()
print("Total Expense: ₹", total_expense)

Total Expense: ₹ 1678901.0


In [ ]:
#Net Savings
net_savings = total_income - total_expense

print("Net Savings: ₹", net_savings)

Net Savings: ₹ -1169127.0


In [ ]:
#Saving Rate
savings_rate = (net_savings / total_income) * 100
print("Savings Rate:", round(savings_rate, 2), "%")

Savings Rate: -229.34 %


In [ ]:
#Top 10 Vendors
top_vendors = (
    df_clean[df_clean["Type"] == "debit"]
    .groupby("Vendor")["Amount"]
    .sum()
    .sort_values(ascending=False)
    .head(5)
)

print(top_vendors)

Vendor
Amazon      328530.0
Zerodha     210000.0
Flipkart    177510.0
Rent        108000.0
Swiggy      104351.0
Name: Amount, dtype: float64


In [ ]:
#Top Spending Categories
category_spending = (
    df_clean[df_clean["Type"] == "debit"]
    .groupby("category")["Amount"]
    .sum()
    .sort_values(ascending=False)
)

print(category_spending)

category
E-commerce           603877.0
Investments          248160.0
Food Delivery        159667.0
Restaurants          117737.0
Rent                 108000.0
Fuel                  89303.0
Groceries             64309.0
Quick Commerce        60152.0
Transport             57474.0
Cash Withdrawal       45500.0
Utilities             41914.0
Cafe                  31445.0
Personal Transfer     24599.0
Subscriptions         18471.0
Entertainment          8293.0
Name: Amount, dtype: float64


In [ ]:
print("Total Transactions:", len(df_clean))

Total Transactions: 1310


## Feature 5 - Monthly Trend Analysis

In [ ]:
#Monthly Spending
df_clean["Month"] = df_clean["Date"].dt.month_name()

In [ ]:
monthly_spending = (
    df_clean[df_clean["Type"]=="debit"]
    .groupby("Month")["Amount"]
    .sum()
)

print(monthly_spending)

Month
April       251382.0
February    234071.0
January     290767.0
June        294917.0
March       330458.0
May         277306.0
Name: Amount, dtype: float64


In [ ]:
df_clean["Month"] = df_clean["Date"].dt.month_name()

monthly_category_spending = (
    df_clean[df_clean["Type"] == "debit"]
    .pivot_table(
        values="Amount",
        index="category",
        columns="Month",
        aggfunc="sum",
        fill_value=0
    )
)

month_order = ["January", "February", "March", "April", "May", "June"]

monthly_category_spending = monthly_category_spending.reindex(
    columns=month_order,
    fill_value=0
)

print(monthly_category_spending)

Month              January  February     March    April      May      June
category                                                                  
Cafe                3690.0    4273.0    5448.0   6564.0   5668.0    5802.0
Cash Withdrawal     2000.0    5000.0    8000.0   5500.0   8000.0   17000.0
E-commerce         98623.0   94011.0  108215.0  69219.0  95776.0  138033.0
Entertainment       1263.0     474.0    2418.0   2224.0      0.0    1914.0
Food Delivery      23692.0   26458.0   25918.0  30092.0  26598.0   26909.0
Fuel               30322.0    2079.0   26164.0  18718.0   9138.0    2882.0
Groceries          17649.0    8571.0    6971.0  13061.0  10336.0    7721.0
Investments        38432.0   15000.0   68644.0  54126.0  48628.0   23330.0
Personal Transfer   7852.0    4285.0    4275.0    663.0   3412.0    4112.0
Quick Commerce      9995.0   12459.0   12229.0   8221.0  10139.0    7109.0
Rent               18000.0   18000.0   18000.0  18000.0  18000.0   18000.0
Restaurants        16320.

In [ ]:
monthly_growth = (
    monthly_category_spending.pct_change(axis=1) * 100
)

print(monthly_growth)

Month              January    February        March      April         May  \
category                                                                     
Cafe                   NaN   15.799458    27.498245  20.484581  -13.650213   
Cash Withdrawal        NaN  150.000000    60.000000 -31.250000   45.454545   
E-commerce             NaN   -4.676394    15.108870 -36.035670   38.366633   
Entertainment          NaN  -62.470309   410.126582  -8.023160 -100.000000   
Food Delivery          NaN   11.674827    -2.040971  16.104638  -11.611059   
Fuel                   NaN  -93.143592  1158.489658 -28.458951  -51.180682   
Groceries              NaN  -51.436342   -18.667600  87.361928  -20.863640   
Investments            NaN  -60.970025   357.626667 -21.149700  -10.157780   
Personal Transfer      NaN  -45.427916    -0.233372 -84.491228  414.630468   
Quick Commerce         NaN   24.652326    -1.846055 -32.774552   23.330495   
Rent                   NaN    0.000000     0.000000   0.000000  

In [ ]:
growth = monthly_growth.replace([float("inf"), -float("inf")], float("nan"))

print("Biggest Month-on-Month Increase:")
print(growth.stack().idxmax(), growth.stack().max())

print("\nBiggest Month-on-Month Decrease:")
print(growth.stack().idxmin(), growth.stack().min())

Biggest Month-on-Month Increase:
('Fuel', 'March') 1158.4896584896585

Biggest Month-on-Month Decrease:
('Entertainment', 'May') -100.0


In [ ]:
# Monthly spending summary for final report

monthly_spend = (
    df_clean[df_clean["Type"] == "debit"]
    .groupby(df_clean[df_clean["Type"] == "debit"]["Date"].dt.strftime("%b"))["Amount"]
    .sum()
)

month_order = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
               "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

monthly_spend = monthly_spend.reindex(
    [m for m in month_order if m in monthly_spend.index]
)

In [ ]:
print("TOP VENDORS BY SPENDING")
print("=" * 60)

max_amount = top_vendors.max()

for vendor, amount in top_vendors.items():
    bar_length = int((amount / max_amount) * 40)  # scale to 40 chars max
    bar = "#" * bar_length
    print(f"{vendor:<15} {bar:<40} Rs. {amount:,.0f}")

TOP VENDORS BY SPENDING
Amazon          ######################################## Rs. 328,530
Zerodha         #########################                Rs. 210,000
Flipkart        #####################                    Rs. 177,510
Rent            #############                            Rs. 108,000
Swiggy          ############                             Rs. 104,351


In [ ]:
print("CATEGORY-WISE SPENDING")
print("=" * 60)

max_amount = category_spending.max()

for category, amount in category_spending.items():
    bar_length = int((amount / max_amount) * 40)
    bar = "#" * bar_length
    print(f"{category:<18} {bar:<40} Rs. {amount:,.0f}")

CATEGORY-WISE SPENDING
E-commerce         ######################################## Rs. 603,877
Investments        ################                         Rs. 248,160
Food Delivery      ##########                               Rs. 159,667
Restaurants        #######                                  Rs. 117,737
Rent               #######                                  Rs. 108,000
Fuel               #####                                    Rs. 89,303
Groceries          ####                                     Rs. 64,309
Quick Commerce     ###                                      Rs. 60,152
Transport          ###                                      Rs. 57,474
Cash Withdrawal    ###                                      Rs. 45,500
Utilities          ##                                       Rs. 41,914
Cafe               ##                                       Rs. 31,445
Personal Transfer  #                                        Rs. 24,599
Subscriptions      #                             

In [ ]:
total = category_spending.sum()

print("EXPENSE DISTRIBUTION BY CATEGORY")
print("=" * 60)

for category, amount in category_spending.sort_values(ascending=False).items():
    pct = (amount / total) * 100
    bar_length = int(pct * 0.8)  # scale so ~100% = 80 chars
    bar = "#" * bar_length
    print(f"{category:<18} {bar:<25} {pct:>5.1f}%   Rs. {amount:,.0f}")

EXPENSE DISTRIBUTION BY CATEGORY
E-commerce         ############################  36.0%   Rs. 603,877
Investments        ###########                14.8%   Rs. 248,160
Food Delivery      #######                     9.5%   Rs. 159,667
Restaurants        #####                       7.0%   Rs. 117,737
Rent               #####                       6.4%   Rs. 108,000
Fuel               ####                        5.3%   Rs. 89,303
Groceries          ###                         3.8%   Rs. 64,309
Quick Commerce     ##                          3.6%   Rs. 60,152
Transport          ##                          3.4%   Rs. 57,474
Cash Withdrawal    ##                          2.7%   Rs. 45,500
Utilities          #                           2.5%   Rs. 41,914
Cafe               #                           1.9%   Rs. 31,445
Personal Transfer  #                           1.5%   Rs. 24,599
Subscriptions                                  1.1%   Rs. 18,471
Entertainment                                  0.

# Feature 6 - Time-of-Day Patterns

In [ ]:
df_clean["Hour"] = pd.to_datetime(df_clean["Time"], format="%H:%M").dt.hour

1.  Hour Extraction

In [ ]:
hourly_spending = (
    df_clean[df_clean["Type"]=="debit"]
    .groupby("Hour")["Amount"]
    .sum()
)

print(hourly_spending)

Hour
0      37969.0
1      38882.0
2      20283.0
3      34868.0
4      43901.0
5      33671.0
6      55506.0
7      19938.0
8      41399.0
9      86701.0
10    140526.0
11    123389.0
12     77096.0
13    113819.0
14    121010.0
15     39647.0
16    100597.0
17     56604.0
18    101343.0
19     67848.0
20    138089.0
21     81545.0
22     55867.0
23     48403.0
Name: Amount, dtype: float64


### Hourly Spending Analysis

In [ ]:
def get_time_period(hour):

    if 5 <= hour < 12:
        return "Morning"

    elif 12 <= hour < 17:
        return "Afternoon"

    elif 17 <= hour < 21:
        return "Evening"

    else:
        return "Night"

df_clean["TimePeriod"] = df_clean["Hour"].apply(get_time_period)

In [ ]:
time_period_spending = (
    df_clean[df_clean["Type"]=="debit"]
    .groupby("TimePeriod")["Amount"]
    .sum()
)

print(time_period_spending)

TimePeriod
Afternoon    452169.0
Evening      363884.0
Morning      501130.0
Night        361718.0
Name: Amount, dtype: float64


PEAK HOUR



In [ ]:
peak_hour = hourly_spending.idxmax()

peak_amount = hourly_spending.max()

print("Peak Spending Hour :", peak_hour)

print("Amount Spent :", peak_amount)

Peak Spending Hour : 10
Amount Spent : 140526.0


In [ ]:
food_delivery = df_clean[
    (df_clean["Type"] == "debit") &
    (df_clean["category"] == "Food Delivery")
]

late_night_food = food_delivery[
    food_delivery["Hour"].isin([21, 22, 23, 0, 1, 2])
]

late_night_percentage = (
    len(late_night_food) / len(food_delivery)
) * 100

print("Late-night Food Delivery transactions:", len(late_night_food))
print("Total Food Delivery transactions:", len(food_delivery))
print("Late-night Food Delivery percentage:", round(late_night_percentage, 2), "%")

Late-night Food Delivery transactions: 78
Total Food Delivery transactions: 364
Late-night Food Delivery percentage: 21.43 %


In [ ]:
df_clean["Hour"] = df_clean["Time"].str[:2].astype(int)

print("\nTIME-OF-DAY PATTERNS")
print("=" * 60)

debits = df_clean[df_clean["Type"] == "debit"]

for cat in debits["category"].unique():
    cat_data = debits[debits["category"] == cat]
    if len(cat_data) == 0:
        continue
    peak_hour = cat_data["Hour"].mode()[0]
    peak_count = (cat_data["Hour"] == peak_hour).sum()
    peak_pct = peak_count / len(cat_data) * 100
    print(f"{cat:<18} peak hour: {peak_hour:02d}:00  ({peak_pct:.0f}% of orders at this hour)")


TIME-OF-DAY PATTERNS
E-commerce         peak hour: 11:00  (6% of orders at this hour)
Transport          peak hour: 20:00  (9% of orders at this hour)
Personal Transfer  peak hour: 17:00  (17% of orders at this hour)
Quick Commerce     peak hour: 19:00  (12% of orders at this hour)
Food Delivery      peak hour: 20:00  (12% of orders at this hour)
Investments        peak hour: 10:00  (26% of orders at this hour)
Entertainment      peak hour: 08:00  (15% of orders at this hour)
Cafe               peak hour: 10:00  (13% of orders at this hour)
Restaurants        peak hour: 20:00  (16% of orders at this hour)
Utilities          peak hour: 12:00  (9% of orders at this hour)
Rent               peak hour: 13:00  (50% of orders at this hour)
Groceries          peak hour: 03:00  (11% of orders at this hour)
Fuel               peak hour: 12:00  (11% of orders at this hour)
Subscriptions      peak hour: 04:00  (26% of orders at this hour)
Cash Withdrawal    peak hour: 09:00  (18% of orders at th



---



# Feature 7 : Anomaly Detection

In [ ]:
df_clean["categoryMean"] = (
    df_clean.groupby("category")["Amount"]
    .transform("mean")
)

In [ ]:
df_clean["categoryStd"] = (
    df_clean.groupby("category")["Amount"]
    .transform("std")
)

In [ ]:
df_clean["ZScore"] = (
    (df_clean["Amount"] - df_clean["categoryMean"])
    /
    df_clean["categoryStd"]
)

In [ ]:
anomalies = df_clean[
    df_clean["ZScore"] > 2
]

anomalies[
    [
        "Date",
        "Vendor",
        "category",
        "Amount",
        "ZScore"
    ]
]

,Date,Vendor,category,Amount,ZScore
49,2024-01-07,Restaurant,Restaurants,7314.0,3.271260
169,2024-01-24,Flipkart,E-commerce,16260.0,2.819266
171,2024-01-24,Indian Oil,Fuel,10408.0,2.246441
216,2024-01-31,DMart,Groceries,2774.0,2.178040
223,2024-02-01,Swiggy,Food Delivery,791.0,2.283167
226,2024-02-02,DMart,Groceries,2704.0,2.081733
269,2024-02-07,Amazon,E-commerce,21986.0,4.085484
336,2024-02-17,Amazon,E-commerce,13674.0,2.247412
414,2024-02-26,Restaurant,Restaurants,8383.0,3.884639
450,2024-03-02,Amazon,E-commerce,18273.0,3.264410


In [ ]:
print("Total Anomalies Found :", len(anomalies))

Total Anomalies Found : 31


4. Top Anomalous Transactions

In [ ]:
top_anomalies = anomalies.sort_values(
    by="Amount",
    ascending=False
)

top_anomalies.head(10)

,Date,Time,Description,Type,Amount,Balance,Mode,Ref,Vendor,vendor_clean,category,Month,Hour,TimePeriod,categoryMean,categoryStd,ZScore
1298,2024-06-26,09:07,AMAZONIN MARKETPLACE,debit,22008.0,-699028.0,UPI,TXN313051,Amazon,Amazon,E-commerce,June,9,Morning,3510.912791,4522.129715,4.090349
269,2024-02-07,17:28,UPI-AMAZONPAY@HDFCBANK,debit,21986.0,-360151.0,UPI,TXN988380,Amazon,Amazon,E-commerce,February,17,Evening,3510.912791,4522.129715,4.085484
475,2024-03-05,21:33,UPI-AMAZONPAY@HDFCBANK,debit,19917.0,-465439.0,UPI,TXN674071,Amazon,Amazon,E-commerce,March,21,Night,3510.912791,4522.129715,3.627956
450,2024-03-02,11:58,AMAZON IN,debit,18273.0,-558650.0,UPI,TXN228115,Amazon,Amazon,E-commerce,March,11,Morning,3510.912791,4522.129715,3.264410
1079,2024-05-27,14:46,UPI-FLIPKART@HDFCBANK,debit,17831.0,-437591.0,UPI,TXN685380,Flipkart,Flipkart,E-commerce,May,14,Afternoon,3510.912791,4522.129715,3.166669
1291,2024-06-25,14:21,Amazon Pay India,debit,17504.0,-629345.0,UPI,TXN103223,Amazon,Amazon,E-commerce,June,14,Afternoon,3510.912791,4522.129715,3.094358
1130,2024-06-03,16:09,UPI-FLIPKART@HDFCBANK,debit,16369.0,-665455.0,UPI,TXN998920,Flipkart,Flipkart,E-commerce,June,16,Afternoon,3510.912791,4522.129715,2.843370
169,2024-01-24,18:43,FLIPKART INDIA,debit,16260.0,-583293.0,UPI,TXN858498,Flipkart,Flipkart,E-commerce,January,18,Evening,3510.912791,4522.129715,2.819266
1308,2024-06-27,16:33,FLIPKART INDIA,debit,15955.0,-540377.0,UPI,TXN403768,Flipkart,Flipkart,E-commerce,June,16,Afternoon,3510.912791,4522.129715,2.751820
651,2024-03-31,13:24,AMAZONIN MARKETPLACE,debit,15953.0,-394113.0,UPI,TXN255130,Amazon,Amazon,E-commerce,March,13,Afternoon,3510.912791,4522.129715,2.751378




---



## Feature 8 - Spending Archetype Detection

In [ ]:
# Category-wise Spending Percentage

category_total = (
    df_clean[df_clean["Type"] == "debit"]
    .groupby("category")["Amount"]
    .sum()
)

print(category_total)

category
Cafe                  31445.0
Cash Withdrawal       45500.0
E-commerce           603877.0
Entertainment          8293.0
Food Delivery        159667.0
Fuel                  89303.0
Groceries             64309.0
Investments          248160.0
Personal Transfer     24599.0
Quick Commerce        60152.0
Rent                 108000.0
Restaurants          117737.0
Subscriptions         18471.0
Transport             57474.0
Utilities             41914.0
Name: Amount, dtype: float64


In [ ]:
category_percent = (
    category_total / category_total.sum()
) * 100

print(category_percent.round(2))

category
Cafe                  1.87
Cash Withdrawal       2.71
E-commerce           35.97
Entertainment         0.49
Food Delivery         9.51
Fuel                  5.32
Groceries             3.83
Investments          14.78
Personal Transfer     1.47
Quick Commerce        3.58
Rent                  6.43
Restaurants           7.01
Subscriptions         1.10
Transport             3.42
Utilities             2.50
Name: Amount, dtype: float64


In [ ]:
# Spending Archetype

def detect_archetypes(df):
    archetypes = []

    total_debit = df.loc[df["Type"] == "debit", "Amount"].sum()

    # Category spending
    category_spend = (
        df[df["Type"] == "debit"]
        .groupby("category")["Amount"]
        .sum()
    )

    # Vendor spending
    vendor_spend = (
        df[df["Type"] == "debit"]
        .groupby("Vendor")["Amount"]
        .sum()
    )

    # Transaction counts
    category_count = (
        df[df["Type"] == "debit"]
        .groupby("category")
        .size()
    )

    # 1. FOODIE
    food_spend = category_spend.get("Food Delivery", 0)
    food_pct = (food_spend / total_debit * 100) if total_debit else 0

    if food_pct >= 15:
        archetypes.append("Foodie")

    # 2. QUICK COMMERCE JUNKIE
    quick_spend = category_spend.get("Quick Commerce", 0)
    quick_pct = (quick_spend / total_debit * 100) if total_debit else 0

    if quick_pct >= 10:
        archetypes.append("Quick Commerce Junkie")

    # 3. SHOPAHOLIC
    ecommerce_spend = category_spend.get("E-commerce", 0)
    ecommerce_pct = (ecommerce_spend / total_debit * 100) if total_debit else 0

    if ecommerce_pct >= 15:
        archetypes.append("Shopaholic")

    # 4. INVESTOR
    investment_spend = category_spend.get("Investments", 0)
    investment_pct = (investment_spend / total_debit * 100) if total_debit else 0

    if investment_pct >= 10:
        archetypes.append("Investor")

    # 5. LATE-NIGHT SNACKER
    food_df = df[
        (df["Type"] == "debit") &
        (df["category"] == "Food Delivery")
    ].copy()

    if len(food_df) > 0:
        late_night = food_df[
            (food_df["Hour"] >= 21) | (food_df["Hour"] <= 1)
        ]

        late_pct = len(late_night) / len(food_df) * 100

        if late_pct >= 40:
            archetypes.append("Late-Night Snacker")

    # 6. CAB COMMUTER
    transport_spend = category_spend.get("Transport", 0)
    transport_pct = (transport_spend / total_debit * 100) if total_debit else 0

    if transport_pct >= 8:
        archetypes.append("Cab Commuter")

    # 7. SUBSCRIPTION LOVER
    subscription_count = category_count.get("Subscriptions", 0)

    if subscription_count >= 5:
        archetypes.append("Subscription Lover")

    # 8. YOLO SPENDER
    savings_rate = (
        (df.loc[df["Type"] == "credit", "Amount"].sum() - total_debit)
        / df.loc[df["Type"] == "credit", "Amount"].sum() * 100
        if df.loc[df["Type"] == "credit", "Amount"].sum() > 0
        else 0
    )

    if savings_rate < 0:
        archetypes.append("YOLO Spender")

    # 9. DISCIPLINED SAVER
    if savings_rate >= 20:
        archetypes.append("Disciplined Saver")

    return archetypes


archetypes = detect_archetypes(df_clean)

print("\n" + "=" * 65)
print("SPENDING ARCHETYPES")
print("=" * 65)

if archetypes:
    for archetype in archetypes:
        print(f"✓ {archetype}")
else:
    print("No strong archetype detected.")


SPENDING ARCHETYPES
✓ Shopaholic
✓ Investor
✓ Subscription Lover
✓ YOLO Spender


In [ ]:
# Late Night Spending Analysis

late_night = df_clean[
    (df_clean["Type"] == "debit") &
    (df_clean["Hour"] >= 22)
]

print("Late Night Transactions:", len(late_night))
print("Late Night Spending: ₹", late_night["Amount"].sum())

Late Night Transactions: 90
Late Night Spending: ₹ 104270.0


In [ ]:
# Favourite Vendor and Category

favorite_category = category_total.idxmax()

print("Most Preferred Spending Category:", favorite_category)

Most Preferred Spending Category: E-commerce


In [ ]:
favorite_vendor = (
    df_clean[df_clean["Type"] == "debit"]
    .groupby("Vendor")["Amount"]
    .sum()
    .idxmax()
)

print("Most Preferred Vendor:", favorite_vendor)

Most Preferred Vendor: Amazon


In [ ]:
# ================================================================
#                    SpendDNA FINAL REPORT
# ================================================================

name = "RAHUL SHARMA"

start_date = df_clean["Date"].min()
end_date = df_clean["Date"].max()

months = df_clean["Date"].dt.to_period("M").nunique()
transactions = len(df_clean)

print("\n" + "=" * 65)
print(f"              SpendDNA REPORT - {name}")
print("\n"
      f"{months} months - "
      f"{transactions:,} transactions - "
      f"{start_date.strftime('%b')} to {end_date.strftime('%b %Y')}"
)
print("\n" + "=" * 65)


# ================================================================
# EXECUTIVE SUMMARY
# ================================================================

total_income = df_clean.loc[
    df_clean["Type"].str.lower() == "credit", "Amount"
].sum()

total_spending = df_clean.loc[
    df_clean["Type"].str.lower() == "debit", "Amount"
].sum()

net_change = total_income - total_spending

savings_rate = (
    (net_change / total_income) * 100
    if total_income > 0 else 0
)

print("\nEXECUTIVE SUMMARY")
print("-" * 65)

print(f"Total credits    : ₹{total_income:,.0f}")
print(f"Total debits     : ₹{total_spending:,.0f}")

if net_change >= 0:
    print(f"Net change       : ₹{net_change:,.0f} (savings)")
else:
    print(f"Net change       : -₹{abs(net_change):,.0f} (overspending)")

if savings_rate >= 0:
    print(f"Savings rate     : {savings_rate:.1f}%")
else:
    print(f"Savings rate     : {savings_rate:.1f}% (BURNING SAVINGS)")

print(f"Transactions     : {len(df_clean):,}")
print(f"Unique vendors   : {df_clean['Vendor'].nunique():,}")


# ================================================================
# TOP SPENDING CATEGORIES
# ================================================================

category_spend = (
    df_clean[
        df_clean["Type"].str.lower() == "debit"
    ]
    .groupby("category")["Amount"]
    .sum()
    .sort_values(ascending=False)
)

print("\nTOP CATEGORIES (% of debit total)")
print("-" * 65)

if len(category_spend) > 0:

    max_category = category_spend.max()

    for category, amount in category_spend.head(5).items():

        percentage = (
            amount / total_spending * 100
            if total_spending > 0 else 0
        )

        bar_length = (
            int((amount / max_category) * 20)
            if max_category > 0 else 0
        )

        print(
            f"{category:<22} "
            f"{'#' * bar_length:<20} "
            f"{percentage:>5.1f}% "
            f"₹{amount:>10,.0f}"
        )


# ================================================================
# TOP SPENDING VENDORS
# ================================================================

vendor_data = df_clean[
    df_clean["Type"].str.lower() == "debit"
]

vendor_spend = (
    vendor_data
    .groupby("Vendor")["Amount"]
    .sum()
    .sort_values(ascending=False)
)

vendor_transactions = (
    vendor_data
    .groupby("Vendor")
    .size()
)

print("\nTOP VENDORS")
print("-" * 65)

for vendor, amount in vendor_spend.head(5).items():

    transactions = vendor_transactions[vendor]

    print(
        f"{vendor:<22} "
        f"₹{amount:>10,.0f} "
        f"({transactions:,} transactions)"
    )


# ================================================================
# TIME-OF-DAY PATTERNS
# ================================================================

print("\nTIME-OF-DAY PATTERNS")
print("-" * 65)

time_spend = (
    vendor_data
    .groupby("TimePeriod")["Amount"]
    .sum()
    .sort_values(ascending=False)
)

for period, amount in time_spend.items():

    percentage = (
        amount / total_spending * 100
        if total_spending > 0 else 0
    )

    print(
        f"{period:<18} "
        f"₹{amount:>10,.0f} "
        f"({percentage:.1f}%)"
    )

if len(time_spend) > 0:
    print(
        f"\nPeak spending period: "
        f"{time_spend.idxmax()}"
    )


# ================================================================
# TOP CATEGORY MONTHLY TREND
# ================================================================

print("\nMONTHLY TREND")
print("-" * 65)

if len(category_spend) > 0:

    top_category = category_spend.index[0]

    top_category_data = vendor_data[
        vendor_data["category"] == top_category
    ].copy()

    top_category_data["Month"] = (
        top_category_data["Date"]
        .dt.strftime("%b")
    )

    monthly_spend = (
        top_category_data
        .groupby("Month")["Amount"]
        .sum()
    )

    month_order = [
        "Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
    ]

    monthly_spend = monthly_spend.reindex(
        [m for m in month_order if m in monthly_spend.index]
    )

    print(f"{top_category} spending:")

    if len(monthly_spend) > 0:

        max_monthly = monthly_spend.max()

        for month, amount in monthly_spend.items():

            bar_length = (
                int((amount / max_monthly) * 20)
                if max_monthly > 0 else 0
            )

            print(
                f"{month:<5} "
                f"₹{amount:>10,.0f} "
                f"{'#' * bar_length}"
            )


# ================================================================
# TOP ANOMALIES
# ================================================================

print("\nTOP ANOMALIES")
print("-" * 65)

anomaly_count = len(anomalies)

print(f"Total anomalous transactions: {anomaly_count}")

if anomaly_count > 0:

    # Sort by absolute z-score if available
    if "z_score" in anomalies.columns:

        top_anomalies = anomalies.copy()

        top_anomalies["abs_z"] = (
            top_anomalies["z_score"].abs()
        )

        top_anomalies = (
            top_anomalies
            .sort_values("abs_z", ascending=False)
            .head(5)
        )

        for _, row in top_anomalies.iterrows():

            date = row["Date"]
            vendor = row["Vendor"]
            amount = row["Amount"]
            z_score = row["z_score"]

            print(
                f"{date.strftime('%d %b')} - "
                f"{vendor} "
                f"₹{amount:,.0f} "
                f"(z={z_score:.1f})"
            )

    else:

        top_anomalies = (
            anomalies
            .sort_values("Amount", ascending=False)
            .head(5)
        )

        for _, row in top_anomalies.iterrows():

            print(
                f"{row['Date'].strftime('%d %b')} - "
                f"{row['Vendor']} "
                f"₹{row['Amount']:,.0f}"
            )


# ================================================================
# SPENDING ARCHETYPES
# ================================================================

print("\nSPENDING ARCHETYPES")
print("-" * 65)

if archetypes:

    for archetype in archetypes:
        print(f"-> {archetype}")

else:

    print("No strong archetype detected.")


# ================================================================
# KEY INSIGHTS
# ================================================================

print("\nKEY INSIGHTS")
print("-" * 65)


# ------------------------------------------------
# Insight 1 - Spending vs Income
# ------------------------------------------------

if total_income > 0:

    spending_percentage = (
        total_spending / total_income
    ) * 100

    if total_spending > total_income:

        monthly_status = "overspending"

        print(
            f"1. Spending is {spending_percentage:.1f}% "
            f"of total income, indicating {monthly_status}."
        )

    else:

        print(
            f"1. Spending is {spending_percentage:.1f}% "
            f"of total income, leaving room for savings."
        )


# ------------------------------------------------
# Insight 2 - Highest Category
# ------------------------------------------------

if len(category_spend) > 0:

    top_category = category_spend.index[0]
    top_category_amount = category_spend.iloc[0]

    top_category_percentage = (
        top_category_amount / total_spending * 100
        if total_spending > 0 else 0
    )

    print(
        f"2. {top_category} is the highest spending "
        f"category, accounting for "
        f"₹{top_category_amount:,.0f} "
        f"({top_category_percentage:.1f}% of total spending)."
    )

# ------------------------------------------------
# Insight 3 - Top Vendor
# ------------------------------------------------

if len(vendor_spend) > 0:

    top_vendor = vendor_spend.index[0]
    top_vendor_amount = vendor_spend.iloc[0]

    print(
        f"3. {top_vendor} is the highest-spend vendor, "
        f"with ₹{top_vendor_amount:,.0f} spent across "
        f"{vendor_transactions[top_vendor]:,} transactions."
    )

# ------------------------------------------------
# Insight 4 - Peak Month
# ------------------------------------------------

if len(monthly_spend) > 0:

    highest_month = monthly_spend.idxmax()
    highest_month_amount = monthly_spend.max()

    print(
        f"4. {highest_month} recorded the highest "
        f"{top_category} spending at "
        f"₹{highest_month_amount:,.0f}."
    )


# ------------------------------------------------
# Insight 5 - Peak Time
# ------------------------------------------------

if len(time_spend) > 0:

    peak_period = time_spend.idxmax()
    peak_period_amount = time_spend.max()

    peak_percentage = (
        peak_period_amount / total_spending * 100
        if total_spending > 0 else 0
    )

    print(
        f"5. The {peak_period} period has the highest "
        f"spending, accounting for "
        f"{peak_percentage:.1f}% of total spending."
    )


# ================================================================
# FINAL STATUS
# ================================================================

print("\n" + "=" * 65)

if net_change < 0:
    print("              SPENDING > INCOME")
else:
    print("              INCOME > SPENDING")
    print("              ✓ POSITIVE CASH FLOW")

print("=" * 65)

print("\n                    END OF REPORT")
print("=" * 65)


              SpendDNA REPORT - RAHUL SHARMA

6 months - 1,310 transactions - Jan to Jun 2024


EXECUTIVE SUMMARY
-----------------------------------------------------------------
Total credits    : ₹509,774
Total debits     : ₹1,678,901
Net change       : -₹1,169,127 (overspending)
Savings rate     : -229.3% (BURNING SAVINGS)
Transactions     : 1,310
Unique vendors   : 40

TOP CATEGORIES (% of debit total)
-----------------------------------------------------------------
E-commerce             ####################  36.0% ₹   603,877
Investments            ########              14.8% ₹   248,160
Food Delivery          #####                  9.5% ₹   159,667
Restaurants            ###                    7.0% ₹   117,737
Rent                   ###                    6.4% ₹   108,000

TOP VENDORS
-----------------------------------------------------------------
Amazon                 ₹   328,530 (86 transactions)
Zerodha                ₹   210,000 (14 transactions)
Flipkart              

## Final Reflection

Building SpendDNA taught me that "messy real-world data" isn't just a warning in a project brief — it's something you only really understand after your code breaks on it.

The hardest part wasn't writing the vendor dictionary or the category mapping — those were mechanical once I inspected the unique descriptions. The real challenge was the date column. My first attempt at `pd.to_datetime()` looked like it worked (no errors thrown), but it silently parsed only 143 out of 1,310 rows correctly. Pandas infers a single date format from the first few rows and quietly fails on everything that doesn't match — even with `errors='coerce'`. I didn't catch this until my monthly summary showed spending in July through December, which was impossible since the dataset only covers January to June. That single bug cascaded through every downstream feature: monthly trends, salary detection, savings rate — all wrong, and none of them threw an error to warn me.

The fix was to parse each of the four known date formats explicitly, one at a time, only applying each format to rows that hadn't already been parsed. It's a good reminder that in data work, the absence of an error message doesn't mean the output is correct — I have to sanity-check the *shape* of my results, not just whether the code ran.

I also learned a smaller but important lesson about consistency: I created near-duplicate columns (`Vendor` and `vendor`, `category` and `Category`) while iterating, which caused several `KeyError`s later. Sticking to one lowercase naming convention from the start, and rebuilding cleanly instead of patching in new cells, would have saved me hours of confusion.

Rahul's numbers confirmed what the brief predicted — he's a Foodie, a Quick Commerce user, an Investor, and a Late-Night Snacker running a deeply negative savings rate. Seeing that pattern emerge from raw, messy transaction strings — not from a clean pre-labeled dataset — is what made this feel like real analytics work rather than a toy exercise.


## Conclusion

SpendDNA successfully cleaned and analyzed Rahul's 6 months of UPI transactions, revealing clear spending patterns — heavy Food Delivery and Quick Commerce usage, healthy investments, but an overall negative savings rate. The project shows how raw, messy bank data can be turned into meaningful financial insights using only core Python, NumPy, and Pandas.